src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -447,7 +447,30 @@ internal void CompleteWriter(Exception? exception)
 
         internal void AdvanceReader(in SequencePosition consumed)
         {
-            AdvanceReader(consumed, consumed);
+            // If the reader is completed
+            if (_readerCompletion.IsCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoReadingAllowed();
+            }
+
+            long examinedIndex = consumed.GetInteger();
+            BufferSegment? examinedSegment = (BufferSegment?)consumed.GetObject();
+            if (examinedSegment != null &&
+                // Avoid the lock if we're examining the entire segment, don't need to look at the last examined index in that case
+                examinedSegment.Length - examinedIndex > 0)
+            {
+                lock (SyncObj)
+                {
+                    // If the last examined index is further than the consumed pointer, let's use the last examined index
+
+                    // _lastExaminedIndex includes the RunningIndex so we remove that to calculate how many bytes we're examining
+                    examinedIndex = Math.Max(examinedIndex, _lastExaminedIndex - examinedSegment.RunningIndex);
+                }
+            }
+
+            // TODO: Use new SequenceMarshal.TryGetReadOnlySequenceSegment to get the correct data
+            // directly casting only works because the type value in ReadOnlySequenceSegment is 0
```

Comment:
No idea what this comment means, but keeping it as it's in the other AdvanceReader overload too.

Reply:
@davidfowl [added the comment](https://github.com/dotnet/corefx/pull/27596), so he can defend it. I wouldn't repeat ths comment. I'd just as soon remove both comments since this code only deals with [`SequenceType.MultiSegment`](https://github.com/dotnet/runtime/blob/b77b71ee2a70b30ee3f9b0884510b62efc74f5c9/src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.cs#L651), so why bother? We'd have to construct a ROS we don't have anyway to use it since `TryGetReadOnlySequenceSegment` doesn't work on `SequencePosition`.